packages/langchain-graph-retriever/src/langchain_graph_retriever/adapters/astra.py

```diff
@@ -226,3 +228,84 @@ async def _aget_by_id_with_embedding(self, document_id: str) -> Content | None:
             projection=self.vector_store.document_codec.full_projection,
         )
         return self._hit_to_content(hit)
+
+
+    @override
+    def get_adjacent(
+        self,
+        edges: set[Edge],
+        query_embedding: list[float],
+        k: int,
+        filter: dict[str, Any] | None,
+        **kwargs: Any,
+    ) -> Iterable[Content]:
+        filter_values: dict[str, Any] = {}
+        for edge in edges:
+            if isinstance(edge, MetadataEdge):
+                filter_values.setdefault(
+                    f"metadata.{edge.incoming_field}", []).append(edge.value)
+            elif isinstance(edge, IdEdge):
+                filter_values.setdefault("_id", []).append(edge.id)
+
+        ids = filter_values.pop("_id", [])
+
+        filter = self.vector_store.filter_to_query(filter) if filter else {}
+        query = { "$or": [ { k: { "$in": v } } for k, v in filter_values.items() ]}
+        if filter:
+            query = { "$and": [filter, query ]}
+
+        results = self._execute_query(
+            k=k,
+            query_embedding=query_embedding,
+            query=query
+        )
+
+        if ids:
+            result_ids = { c.id for c in results }
+            ids = [ id for id in ids if id not in result_ids ]
+
+            ids_query = {"_id": { "$in": ids }}
+            if filter:
+                ids_query = { "$and": [filter, ids_query] }
+            if ids:
+                results.extend(self._execute_query(
+                    k=k,
+                    query_embedding=query_embedding,
+                    query=ids_query,
+                ))
+        return results
+
+    def _execute_query(self,
```

Comment:
At this point, I don't know that the Astra vector store is providing much. We basically reach in to get the configured astra environment, then issue queries directly against it. It may be worth discussing just moving this to `graph_retriever` and reading directly.

Reply:
I had to do this because going through langchain rewrites the query. I *believe* it doesn't rewrite keys starting with `$` so we'd be able to express the metadata filters, but it doesn't like the `_id` part of the query.

Reply:
I'm surprised to see this code in the *langchain* astra adapter.  I would have expected it at the `graph-retriever` level, since it circumvents nearly all the langchain-astra-db stuff.

Reply:
I think the `$` is related to vectorize.

Reply:
Yeah -- it still uses some bits from langchain (such as the different filter encodings) because sometimes it's flat and sometimes it's not. As noted, I think this may be the point at which we should consider moving it to `graph-retriever`, but we should decide if we want to support all the cases that `langchain` currently does. I'll send some questions to Stefano I think to clairfy.

The `$` is just a way of encoding query operations in the request. 

https://docs.datastax.com/en/astra-db-serverless/api-reference/documents.html#operators

*some* of those operators relate to vectorize.